### ISSUE TITLE
`unnecessary_cast` is removing parens by accident

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#16475]
Using the following flags

`--force-warn clippy::unnecessary_cast`

this code:

```rust
pub fn maybe_ptr() -> *const u8 {
    static NONE: Option<((), &'static u8)> = None;
    unsafe {
        *(&NONE as *const _ as *const _ as *const *const u8 as *const *const u8)
    }
}
```

caused the following diagnostics:

```
    Checking _776b69056990452c4caa934562557fea47695b56 v0.1.0 (/tmp/icemaker_global_tempdir.dESjtcUkYIKG/icemaker_clippyfix_tempdir.46vD76CDfUdM/_776b69056990452c4caa934562557fea47695b56)
warning: casting raw pointers to the same type and constness is unnecessary (`*const *const u8` -> `*const *const u8`)
 --> src/lib.rs:4:10
  |
4 |         *(&NONE as *const _ as *const _ as *const *const u8 as *const *const u8)
  |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `&NONE as *const _ as *const _ as *const *const u8`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_cast
  = note: requested on the command line with `--force-warn clippy::unnecessary-cast`

warning: `_776b69056990452c4caa934562557fea47695b56` (lib) generated 1 warning
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.24s

```

However after applying these diagnostics, the resulting code:

```rust
pub fn maybe_ptr() -> *const u8 {
    static NONE: Option<((), &'static u8)> = None;
    unsafe {
        *&NONE as *const _ as *const _ as *const *const u8
    }
}
```
no longer compiled:

```
    Checking _776b69056990452c4caa934562557fea47695b56 v0.1.0 (/tmp/icemaker_global_tempdir.dESjtcUkYIKG/icemaker_clippyfix_tempdir.46vD76CDfUdM/_776b69056990452c4caa934562557fea47695b56)
error[E0308]: mismatched types
 --> src/lib.rs:4:9
  |
1 | pub fn maybe_ptr() -> *const u8 {
  |                       --------- expected `*const u8` because of return type
...
4 |         *&NONE as *const _ as *const _ as *const *const u8
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `*const u8`, found `*const *const u8`
  |
  = note: expected raw pointer `*const u8`
             found raw pointer `*const *const u8`

For more information about this error, try `rustc --explain E0308`.
error: could not compile `_776b69056990452c4caa934562557fea47695b56` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `_776b69056990452c4caa934562557fea47695b56` (lib test) due to 1 previous error

```

Version:
```
rustc 1.95.0-nightly (94a0cd15f 2026-01-27)
binary: rustc
commit-hash: 94a0cd15f5976fa35e5e6784e621c04e9f958e57
commit-date: 2026-01-27
host: x86_64-unknown-linux-gnu
release: 1.95.0-nightly
LLVM version: 21.1.8

```

**Comment by matthiaskrgr** [comment]
suggeston should be  `*(&NONE as *const _ as *const _ as *const *const u8)` I guess?
